src/Microsoft.Framework.Runtime.Sources/Impl/ApplicationGlobalData.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Framework.Runtime.Sources.Impl
+{
+    // Helper class to completely centralize all of the application global data logic and avoid #if statements elsewhere.
+    internal class ApplicationGlobalData
+    {
+        private readonly IApplicationEnvironment _hostEnvironment;
+#if DNXCORE50
+        private readonly object _lock;
+        private readonly Dictionary<string, object> _appGlobalData;
```

Comment:
In general, we try to expose a consistent facade over whatever the runtime provides. In this case I think it's fine to use AppDomain.CurrentDomain.Get/SetData, similarly to how we use LogicalCallContext to implement certain features on .NET Framework. Last I heard, the CLR team were thinking about adding a general bag of things just like AppDomain.Get/SetData and I think if that were available, we'd use it.

/cc @jkotas